Regression in txpool dump and restore on v25.3.0

### Steps to Reproduce
1. Run Besu on Ethereum mainnet with `--tx-pool=layered --tx-pool-enable-save-restore=true` with no txpool dump file present
2. Leave Besu running for about 5 minutes to saturate the transaction pool and quit Besu
3. Run Besu again with the same arguments

**Expected behavior:** See "Added X transactions of Y loaded from file Z, before operation was cancelled" reported within the logs

**Actual behavior:** An non-fatal error is reported as originating from the component TransactionPool "Error while restoring transaction pool from disk" with a stacktrace

**Frequency:** Almost every single restart of Besu on Ethereum mainnet since updating to v23.5.0

### Logs
```
2025-03-21 04:32:45.062+00:00 | main | INFO  | NetworkRunner | Supported capabilities: [eth/63, eth/64, eth/65, eth/66, eth/67, eth/68], [snap/1]
2025-03-21 04:32:45.062+00:00 | main | INFO  | DefaultSynchronizer | Starting synchronizer.
2025-03-21 04:32:45.073+00:00 | main | INFO  | TransactionPoolFactory | Enabling transaction handling following initial sync
2025-03-21 04:32:45.084+00:00 | ForkJoinPool.commonPool-worker-7 | INFO  | TransactionPool | Loading transaction pool content from file /var/lib/besu/txpool.dump
2025-03-21 04:32:45.085+00:00 | main | INFO  | DefaultSynchronizer | Stopping block propagation.
2025-03-21 04:32:45.085+00:00 | main | INFO  | DefaultSynchronizer | Stopping the pruner.
2025-03-21 04:32:45.086+00:00 | main | INFO  | Runner | Ethereum main loop is up.
2025-03-21 04:32:47.803+00:00 | ForkJoinPool.commonPool-worker-7 | ERROR | TransactionPool | Error while restoring transaction pool from disk
java.util.concurrent.CompletionException: org.hyperledger.besu.ethereum.rlp.RLPException: Cannot read a arbitrary precision scalar, current item is a list (at bytes 4-1348: fa06067a[f9053d01...b4b9b664]fa06000c...)
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1807)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1796)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
Caused by: org.hyperledger.besu.ethereum.rlp.RLPException: Cannot read a arbitrary precision scalar, current item is a list (at bytes 4-1348: fa06067a[f9053d01...b4b9b664]fa06000c...)
	at org.hyperledger.besu.ethereum.rlp.AbstractRLPInput.error(AbstractRLPInput.java:205)
	at org.hyperledger.besu.ethereum.rlp.AbstractRLPInput.checkElt(AbstractRLPInput.java:250)
	at org.hyperledger.besu.ethereum.rlp.AbstractRLPInput.checkScalar(AbstractRLPInput.java:263)
	at org.hyperledger.besu.ethereum.rlp.AbstractRLPInput.readBigIntegerScalar(AbstractRLPInput.java:337)
	at org.hyperledger.besu.ethereum.rlp.BytesValueRLPInput.readBigIntegerScalar(BytesValueRLPInput.java:24)
	at org.hyperledger.besu.ethereum.core.encoding.BlobTransactionDecoder.readTransactionPayloadInner(BlobTransactionDecoder.java:70)
	at org.hyperledger.besu.ethereum.core.encoding.BlobTransactionDecoder.readTransactionPayload(BlobTransactionDecoder.java:53)
	at org.hyperledger.besu.ethereum.core.encoding.BlobTransactionDecoder.decode(BlobTransactionDecoder.java:47)
	at org.hyperledger.besu.ethereum.core.encoding.TransactionDecoder.decodeTypedTransaction(TransactionDecoder.java:102)
	at org.hyperledger.besu.ethereum.core.encoding.TransactionDecoder.decodeTypedTransaction(TransactionDecoder.java:82)
	at org.hyperledger.besu.ethereum.core.encoding.TransactionDecoder.decodeRLP(TransactionDecoder.java:60)
	at org.hyperledger.besu.ethereum.core.Transaction.readFrom(Transaction.java:140)
	at org.hyperledger.besu.ethereum.core.Transaction.readFrom(Transaction.java:136)
	at org.hyperledger.besu.ethereum.eth.transactions.TransactionPool$SaveRestoreManager.lambda$executeLoadFromDisk$6(TransactionPool.java:883)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.WhileOps$1$1.accept(WhileOps.java:99)
	at java.base/java.util.Spliterators$IteratorSpliterator.tryAdvance(Spliterators.java:1950)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.hyperledger.besu.ethereum.eth.transactions.TransactionPool$SaveRestoreManager.executeLoadFromDisk(TransactionPool.java:890)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
	... 6 more
```

### Versions (Add all that apply)
* Software version: `besu/v25.3.0/linux-x86_64/openjdk-java-21` (Docker image)
* Java version:
```
openjdk version "21.0.6" 2025-01-21
OpenJDK Runtime Environment (build 21.0.6+7-Ubuntu-124.04.1)
OpenJDK 64-Bit Server VM (build 21.0.6+7-Ubuntu-124.04.1, mixed mode, sharing)
```
* OS Name & Version:
```
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=24.04
DISTRIB_CODENAME=noble
DISTRIB_DESCRIPTION="Ubuntu 24.04.1 LTS"
PRETTY_NAME="Ubuntu 24.04.1 LTS"
NAME="Ubuntu"
VERSION_ID="24.04"
VERSION="24.04.1 LTS (Noble Numbat)"
VERSION_CODENAME=noble
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=noble
LOGO=ubuntu-logo
```
* Kernel Version: 
```
Linux [redacted] 6.13.7-arch1-1 #1 SMP PREEMPT_DYNAMIC Thu, 13 Mar 2025 18:12:00 +0000 x86_64 x86_64 x86_64 GNU/Linux
```
(Arch Linux as Podman container host)
* Virtual Machine software & version: `N/A`
* Docker Version:
```
Client:       Podman Engine
Version:      5.4.1
API Version:  5.4.1
Go Version:   go1.24.1
Git Commit:   b79bc8afe796cba51dd906270a7e1056ccdfcf9e
Built:        Wed Mar 12 07:07:00 2025
OS/Arch:      linux/amd64
```
* Cloud VM, type, size: `N/A / Bare metal`
* Consensus Client & Version if using Proof of Stake: Teku 25.3.0

### Additional Information
My best guess from poking around, is that this was introduced in [74a1d8a1ea7f... (Support pending transaction score when saving and restoring txpool (#8363))](https://github.com/hyperledger/besu/commit/74a1d8a1ea7f6cf29fee0ee60b260160f0c17473).

This PR attempts to let `org.hyperledger.besu.ethereum.eth.transactions.TransactionPool` make use of the [POOLED_TRANSACTION encoding context when saving the dump file under executeSaveToDisk()](https://github.com/hyperledger/besu/blob/74a1d8a1ea7f6cf29fee0ee60b260160f0c17473/ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/TransactionPool.java#L829). While decoding is still being performed under executeLoadFromDisk() through the `org.hyperledger.besu.ethereum.core.Transaction.readFrom()` wrapper, which [assumes a BLOCK_BODY context](https://github.com/hyperledger/besu/blob/2412586d06c2c0d1424fbfacbbababa62f2a5a74/ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/Transaction.java#L140).

Unrelated, but TransactionPool.executeLoadFromDisk() appears to report the wrong error message [when throwing an IOException](https://github.com/hyperledger/besu/blob/74a1d8a1ea7f6cf29fee0ee60b260160f0c17473/ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/TransactionPool.java#L915).

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -31,6 +31,7 @@
 - Fix QBFT and IBFT transitions that change the mining beneficiary [#8387](https://github.com/hyperledger/besu/issues/8387)
 - `eth_getLogs` - empty topic is a wildcard match [#8420](https://github.com/hyperledger/besu/pull/8420)
 - Upgrade spring-security-crypto to address CVE-2025-22228 [#8448](https://github.com/hyperledger/besu/pull/8448)
+- Fix restoring txpool from disk when blob transactions are present [#8481](https://github.com/hyperledger/besu/pull/8481)
 
 ## 25.3.0 
 

--- a/ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/Transaction.java
+++ b/ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/Transaction.java
@@ -136,8 +136,25 @@ public class Transaction
     return readFrom(RLP.input(rlpBytes));
   }
 
+  /**
+   * Recreate a transaction from RLP serialized in the block body format
+   *
+   * @param rlpInput the RLP input in block body format
+   * @return the transaction
+   */
   public static Transaction readFrom(final RLPInput rlpInput) {
-    return TransactionDecoder.decodeRLP(rlpInput, EncodingContext.BLOCK_BODY);
+    return readFrom(rlpInput, EncodingContext.BLOCK_BODY);
+  }
+
+  /**
+   * Recreate a transaction from RLP serialized according to the specified format
+   *
+   * @param rlpInput the RLP input in block body format
+   * @param context specify in which format the RLP was serialized
+   * @return the transaction
+   */
+  public static Transaction readFrom(final RLPInput rlpInput, final EncodingContext context) {
+    return TransactionDecoder.decodeRLP(rlpInput, context);
   }
 
   /**
@@ -484,12 +501,22 @@ public class Transaction
   }
 
   /**
-   * Writes the transaction to RLP
+   * Writes the transaction to RLP using the block body format
    *
    * @param out the output to write the transaction to
    */
   public void writeTo(final RLPOutput out) {
-    TransactionEncoder.encodeRLP(this, out, EncodingContext.BLOCK_BODY);
+    writeTo(out, EncodingContext.BLOCK_BODY);
+  }
+
+  /**
+   * Writes the transaction to RLP using the specified format
+   *
+   * @param out the output to write the transaction to
+   * @param context the serialization format to use
+   */
+  public void writeTo(final RLPOutput out, final EncodingContext context) {
+    TransactionEncoder.encodeRLP(this, out, context);
   }
 
   @Override

--- a/ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/TransactionPool.java
+++ b/ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/TransactionPool.java
@@ -34,7 +34,6 @@ import org.hyperledger.besu.ethereum.chain.MutableBlockchain;
 import org.hyperledger.besu.ethereum.core.BlockHeader;
 import org.hyperledger.besu.ethereum.core.Transaction;
 import org.hyperledger.besu.ethereum.core.encoding.EncodingContext;
-import org.hyperledger.besu.ethereum.core.encoding.TransactionEncoder;
 import org.hyperledger.besu.ethereum.eth.manager.EthContext;
 import org.hyperledger.besu.ethereum.eth.manager.EthPeer;
 import org.hyperledger.besu.ethereum.eth.manager.EthScheduler;
@@ -44,6 +43,7 @@ import org.hyperledger.besu.ethereum.mainnet.TransactionValidator;
 import org.hyperledger.besu.ethereum.mainnet.ValidationResult;
 import org.hyperledger.besu.ethereum.mainnet.feemarket.FeeMarket;
 import org.hyperledger.besu.ethereum.rlp.BytesValueRLPOutput;
+import org.hyperledger.besu.ethereum.rlp.RLP;
 import org.hyperledger.besu.ethereum.transaction.TransactionInvalidReason;
 import org.hyperledger.besu.ethereum.trie.MerkleTrieException;
 import org.hyperledger.besu.evm.account.Account;
@@ -825,8 +825,7 @@ public class TransactionPool implements BlockAddedObserver {
                 .map(
                     ptx -> {
                       final BytesValueRLPOutput rlp = new BytesValueRLPOutput();
-                      TransactionEncoder.encodeRLP(
-                          ptx.getTransaction(), rlp, EncodingContext.POOLED_TRANSACTION);
+                      ptx.getTransaction().writeTo(rlp, EncodingContext.POOLED_TRANSACTION);
                       return ptx.getScore()
                           + (ptx.isReceivedFromLocalSource() ? "l" : "r")
                           + rlp.encoded().toBase64String();
@@ -881,7 +880,10 @@ public class TransactionPool implements BlockAddedObserver {
                           final boolean isLocal = line.charAt(scoreStr.length()) == 'l';
                           final Transaction tx =
                               Transaction.readFrom(
-                                  Bytes.fromBase64String(line.substring(scoreStr.length() + 1)));
+                                  RLP.input(
+                                      Bytes.fromBase64String(
+                                          line.substring(scoreStr.length() + 1))),
+                                  EncodingContext.POOLED_TRANSACTION);
 
                           final ValidationResult<TransactionInvalidReason> result =
                               addTransaction(tx, isLocal, score);
@@ -912,7 +914,7 @@ public class TransactionPool implements BlockAddedObserver {
               saveFile.delete();
             }
           } catch (IOException e) {
-            LOG.error("Error while saving txpool content to disk", e);
+            LOG.error("Error while loading txpool content from disk", e);
           }
         }
       }

--- a/ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/transactions/AbstractTransactionPoolTest.java
+++ b/ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/transactions/AbstractTransactionPoolTest.java
@@ -25,11 +25,13 @@ import static org.hyperledger.besu.ethereum.transaction.TransactionInvalidReason
 import static org.hyperledger.besu.ethereum.transaction.TransactionInvalidReason.NONCE_TOO_FAR_IN_FUTURE_FOR_SENDER;
 import static org.hyperledger.besu.ethereum.transaction.TransactionInvalidReason.NONCE_TOO_LOW;
 import static org.hyperledger.besu.ethereum.transaction.TransactionInvalidReason.REPLAY_PROTECTED_SIGNATURE_REQUIRED;
+import static org.hyperledger.besu.ethereum.transaction.TransactionInvalidReason.TRANSACTION_ALREADY_KNOWN;
 import static org.hyperledger.besu.ethereum.transaction.TransactionInvalidReason.TRANSACTION_REPLACEMENT_UNDERPRICED;
 import static org.hyperledger.besu.ethereum.transaction.TransactionInvalidReason.TX_FEECAP_EXCEEDED;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.eq;
-import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.atLeastOnce;
+import static org.mockito.Mockito.clearInvocations;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.verifyNoInteractions;
@@ -103,9 +105,9 @@ public abstract class AbstractTransactionPoolTest extends AbstractTransactionPoo
     addAndAssertRemoteTransactionsValid(transaction0);
     addAndAssertRemoteTransactionsValid(transaction1);
 
-    assertThat(transactions.size()).isEqualTo(3);
-    assertThat(transactions.getLocalTransactions()).contains(localTransaction2);
-    assertThat(transactions.getPriorityTransactions()).hasSize(noLocalPriority ? 0 : 1);
+    assertThat(transactionPool.count()).isEqualTo(3);
+    assertThat(getLocalTransactions()).contains(localTransaction2);
+    assertThat(getPriorityTransactions()).hasSize(noLocalPriority ? 0 : 1);
   }
 
   @Test
@@ -131,7 +133,7 @@ public abstract class AbstractTransactionPoolTest extends AbstractTransactionPoo
 
     assertTransactionNotPending(transaction0);
     assertTransactionNotPending(transaction1);
-    assertThat(transactions.size()).isZero();
+    assertThat(transactionPool.count()).isZero();
   }
 
   @Test
@@ -144,7 +146,7 @@ public abstract class AbstractTransactionPoolTest extends AbstractTransactionPoo
 
     assertTransactionNotPending(transaction0);
     assertTransactionNotPending(transaction1);
-    assertThat(transactions.size()).isZero();
+    assertThat(transactionPool.count()).isZero();
   }
 
   @Test
@@ -197,19 +199,19 @@ public abstract class AbstractTransactionPoolTest extends AbstractTransactionPoo
         appendBlock(Difficulty.ONE, originalFork1.getHeader(), transactionOtherSender);
     assertTransactionNotPending(transaction0);
     assertTransactionNotPending(transactionOtherSender);
-    assertThat(transactions.getLocalTransactions()).isEmpty();
+    assertThat(getLocalTransactions()).isEmpty();
 
     final Block reorgFork1 = appendBlock(Difficulty.ONE, commonParent);
     verifyChainHeadIs(originalFork2);
 
-    transactions.subscribePendingTransactions(listener);
+    transactionPool.subscribePendingTransactions(listener);
     final Block reorgFork2 = appendBlock(Difficulty.of(2000), reorgFork1.getHeader());
     verifyChainHeadIs(reorgFork2);
 
     assertTransactionPending(transaction0);
     assertTransactionPending(transactionOtherSender);
-    assertThat(transactions.getLocalTransactions()).contains(transaction0);
-    assertThat(transactions.getLocalTransactions()).doesNotContain(transactionOtherSender);
+    assertThat(getLocalTransactions()).contains(transaction0);
+    assertThat(getLocalTransactions()).doesNotContain(transactionOtherSender);
     verify(listener).onTransactionAdded(transaction0);
     verify(listener).onTransactionAdded(transactionOtherSender);
     verifyNoMoreInteractions(listener);
@@ -275,7 +277,7 @@ public abstract class AbstractTransactionPoolTest extends AbstractTransactionPoo
     assertTransactionPending(transactionWithBlobs);
 
     Optional<Transaction> maybeBlob =
-        transactions.getTransactionByHash(transactionWithBlobs.getHash());
+        transactionPool.getTransactionByHash(transactionWithBlobs.getHash());
     assertThat(maybeBlob).isPresent();
     Transaction restoredBlob = maybeBlob.get();
     assertThat(restoredBlob).isEqualTo(transactionWithBlobs);
@@ -384,10 +386,19 @@ public abstract class AbstractTransactionPoolTest extends AbstractTransactionPoo
 
   @Test
   public void shouldDiscardRemoteTransactionThatAlreadyExistsBeforeValidation() {
-    doReturn(true).when(transactions).containsTransaction(transaction0);
-    transactionPool.addRemoteTransactions(singletonList(transaction0));
+    givenTransactionIsValid(transaction0);
+    addAndAssertRemoteTransactionsValid(transaction0);
 
-    verify(transactions).containsTransaction(transaction0);
+    verify(transactionValidatorFactory, atLeastOnce()).get();
+
+    clearInvocations(transactionValidatorFactory);
+
+    // trying to re-add the same tx should return transaction already known and no
+    // access to the transaction validator factory should be done
+    final var result = transactionPool.addRemoteTransactions(singletonList(transaction0));
+    assertThat(result.get(transaction0.getHash()))
+        .isEqualTo(ValidationResult.invalid(TRANSACTION_ALREADY_KNOWN));
+
     verifyNoInteractions(transactionValidatorFactory);
   }
 

--- a/ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/transactions/AbstractTransactionPoolTestBase.java
+++ b/ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/transactions/AbstractTransactionPoolTestBase.java
@@ -126,7 +126,6 @@ public abstract class AbstractTransactionPoolTestBase {
   protected MutableBlockchain blockchain;
   protected TransactionBroadcaster transactionBroadcaster;
 
-  protected PendingTransactions transactions;
   protected final Transaction transaction0 = createTransaction(0);
   protected final Transaction transaction1 = createTransaction(1);
   protected final Transaction transactionWithBlobs = createBlobTransaction(2);
@@ -270,11 +269,9 @@ public abstract class AbstractTransactionPoolTestBase {
             transactionReplacementHandler.shouldReplace(
                 t1, t2, protocolContext.getBlockchain().getChainHeadHeader());
 
-    transactions = spy(createPendingTransactions(poolConfig, transactionReplacementTester));
-
     final TransactionPool txPool =
         new TransactionPool(
-            () -> transactions,
+            () -> createPendingTransactions(poolConfig, transactionReplacementTester),
             protocolSchedule,
             protocolContext,
             transactionBroadcaster,
@@ -310,7 +307,7 @@ public abstract class AbstractTransactionPoolTestBase {
   }
 
   protected void assertTransactionNotPending(final Transaction transaction) {
-    assertThat(transactions.getTransactionByHash(transaction.getHash())).isEmpty();
+    assertThat(transactionPool.getTransactionByHash(transaction.getHash())).isEmpty();
   }
 
   protected void addAndAssertRemoteTransactionInvalid(final Transaction tx) {
@@ -321,7 +318,7 @@ public abstract class AbstractTransactionPoolTestBase {
   }
 
   protected void assertTransactionPending(final Transaction t) {
-    assertThat(transactions.getTransactionByHash(t.getHash())).contains(t);
+    assertThat(transactionPool.getTransactionByHash(t.getHash())).contains(t);
   }
 
   protected void addAndAssertRemoteTransactionsValid(final Transaction... txs) {
@@ -340,9 +337,9 @@ public abstract class AbstractTransactionPoolTestBase {
         .onTransactionsAdded(
             argThat(btxs -> btxs.size() == txs.length && btxs.containsAll(List.of(txs))));
     Arrays.stream(txs).forEach(this::assertTransactionPending);
-    assertThat(transactions.getLocalTransactions()).doesNotContain(txs);
+    assertThat(getLocalTransactions()).doesNotContain(txs);
     if (hasPriority) {
-      assertThat(transactions.getPriorityTransactions()).contains(txs);
+      assertThat(getPriorityTransactions()).contains(txs);
     }
   }
 
@@ -354,11 +351,11 @@ public abstract class AbstractTransactionPoolTestBase {
     assertThat(result.isValid()).isTrue();
     assertTransactionPending(tx);
     verify(transactionBroadcaster).onTransactionsAdded(singletonList(tx));
-    assertThat(transactions.getLocalTransactions()).contains(tx);
+    assertThat(getLocalTransactions()).contains(tx);
     if (disableLocalPriority) {
-      assertThat(transactions.getPriorityTransactions()).doesNotContain(tx);
+      assertThat(getPriorityTransactions()).doesNotContain(tx);
     } else {
-      assertThat(transactions.getPriorityTransactions()).contains(tx);
+      assertThat(getPriorityTransactions()).contains(tx);
     }
   }
 
@@ -551,7 +548,7 @@ public abstract class AbstractTransactionPoolTestBase {
       transactionPool.addRemoteTransactions(List.of(transaction));
     }
 
-    return transactions.size();
+    return transactionPool.count();
   }
 
   protected Block appendBlockGasPriceMarket(
@@ -597,5 +594,19 @@ public abstract class AbstractTransactionPoolTestBase {
             .collect(toList());
     blockchain.appendBlock(block, transactionReceipts);
     return block;
+  }
+
+  protected List<Transaction> getPriorityTransactions() {
+    return transactionPool.getPendingTransactions().stream()
+        .filter(PendingTransaction::hasPriority)
+        .map(PendingTransaction::getTransaction)
+        .toList();
+  }
+
+  protected List<Transaction> getLocalTransactions() {
+    return transactionPool.getPendingTransactions().stream()
+        .filter(PendingTransaction::isReceivedFromLocalSource)
+        .map(PendingTransaction::getTransaction)
+        .toList();
   }
 }

--- a/ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/transactions/TransactionPoolSaveRestoreTest.java
+++ b/ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/transactions/TransactionPoolSaveRestoreTest.java
@@ -140,6 +140,25 @@ public class TransactionPoolSaveRestoreTest extends AbstractTransactionPoolTestB
   @ValueSource(booleans = {true, false})
   public void localTransactionIsSavedAndRestored(final boolean noLocalPriority)
       throws ExecutionException, InterruptedException, TimeoutException, IOException {
+    transactionIsSavedAndRestored(
+        createTransaction(noLocalPriority ? 0 : 1), true, noLocalPriority);
+  }
+
+  @Test
+  public void remoteTransactionIsSavedAndRestored()
+      throws ExecutionException, InterruptedException, TimeoutException, IOException {
+    transactionIsSavedAndRestored(createTransaction(2), false, true);
+  }
+
+  @Test
+  public void blobTransactionIsSavedAndRestored()
+      throws ExecutionException, InterruptedException, TimeoutException, IOException {
+    transactionIsSavedAndRestored(createBlobTransaction(0), false, true);
+  }
+
+  private void transactionIsSavedAndRestored(
+      final Transaction transaction, final boolean isLocal, final boolean noLocalPriority)
+      throws ExecutionException, InterruptedException, TimeoutException, IOException {
     // create a txpool with save and restore enabled
     this.transactionPool =
         createTransactionPool(
@@ -148,11 +167,13 @@ public class TransactionPoolSaveRestoreTest extends AbstractTransactionPoolTestB
                     .enableSaveRestore(true)
                     .saveFile(saveFilePath.toFile()));
 
-    final Transaction transaction = createTransaction(noLocalPriority ? 0 : 1);
-
     givenTransactionIsValid(transaction);
 
-    addAndAssertTransactionViaApiValid(transaction, noLocalPriority);
+    if (isLocal) {
+      addAndAssertTransactionViaApiValid(transaction, noLocalPriority);
+    } else {
+      addAndAssertRemoteTransactionsValid(transaction);
+    }
 
     // disabling the txpool, forces a save to file
     transactionPool.setDisabled().get(10, TimeUnit.SECONDS);
@@ -162,7 +183,9 @@ public class TransactionPoolSaveRestoreTest extends AbstractTransactionPoolTestB
 
     final var savedContent = Files.readString(saveFilePath, StandardCharsets.US_ASCII);
 
-    assertThat(savedContent).isEqualToIgnoringNewLines("127l" + transaction2Base64(transaction));
+    assertThat(savedContent)
+        .isEqualToIgnoringNewLines(
+            "127" + ((isLocal) ? "l" : "r") + transaction2Base64(transaction));
 
     // re-enabling the txpool restores from file
     transactionPool.setEnabled().get(10, TimeUnit.SECONDS);
@@ -172,42 +195,8 @@ public class TransactionPoolSaveRestoreTest extends AbstractTransactionPoolTestB
     final var restoredPendingTx = transactionPool.getPendingTransactions().iterator().next();
 
     assertThat(restoredPendingTx.getTransaction()).isEqualTo(transaction);
-    assertThat(restoredPendingTx.isReceivedFromLocalSource()).isTrue();
+    assertThat(restoredPendingTx.isReceivedFromLocalSource()).isEqualTo(isLocal);
     assertThat(restoredPendingTx.hasPriority()).isNotEqualTo(noLocalPriority);
-  }
-
-  @Test
-  public void remoteTransactionIsSavedAndRestored()
-      throws ExecutionException, InterruptedException, TimeoutException, IOException {
-    // create a txpool with save and restore enabled
-    this.transactionPool =
-        createTransactionPool(b -> b.enableSaveRestore(true).saveFile(saveFilePath.toFile()));
-
-    final Transaction transaction = createTransaction(2);
-
-    givenTransactionIsValid(transaction);
-
-    addAndAssertRemoteTransactionsValid(transaction);
-
-    // disabling the txpool, forces a save to file
-    transactionPool.setDisabled().get(10, TimeUnit.SECONDS);
-
-    // after being disabled the txpool must be empty
-    assertThat(transactionPool.getPendingTransactions()).isEmpty();
-
-    final var savedContent = Files.readString(saveFilePath, StandardCharsets.US_ASCII);
-
-    assertThat(savedContent).isEqualToIgnoringNewLines("127r" + transaction2Base64(transaction));
-
-    // re-enabling the txpool restores from file
-    transactionPool.setEnabled().get(10, TimeUnit.SECONDS);
-
-    assertThat(transactionPool.getPendingTransactions()).size().isEqualTo(1);
-
-    final var restoredPendingTx = transactionPool.getPendingTransactions().iterator().next();
-
-    assertThat(restoredPendingTx.getTransaction()).isEqualTo(transaction);
-    assertThat(restoredPendingTx.isReceivedFromLocalSource()).isFalse();
   }
 
   @Test

